src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,215 @@
+    assert(_nextVelocity.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 1, _currVelocity);
+    splineProblem.addConstantConstraint(1, 1, _nextVelocity);
+  }
+
+  if (_numCoefficients >= 6)
+  {
+    assert(_currAcceleration.size() == numDofs);
+    assert(_nextAcceleration.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 2, _currAcceleration);
+    splineProblem.addConstantConstraint(1, 2, _nextAcceleration);
+  }
+
+  const auto splineSegment = splineProblem.fit();
+  return splineSegment.getCoefficients()[0];
+}
+
+//=============================================================================
+void extractJointTrajectoryPoint(
+  const trajectory_msgs::JointTrajectory& _trajectory,
+  size_t _index, size_t _numDofs,
+  Eigen::VectorXd* _positions, bool _positionsRequired,
+  Eigen::VectorXd* _velocities, bool _velocitiesRequired,
+  Eigen::VectorXd* _accelerations, bool _accelerationsRequired)
+{
+  const auto& waypoint = _trajectory.points[_index];
+
+  try
+  {
+    checkVector("positions", waypoint.positions, _numDofs,
+      _positionsRequired, _positions);
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, _velocities);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& _space,
+  const trajectory_msgs::JointTrajectory& _jointTrajectory)
+{
+  if (!_space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledDofs = _space->getNumSubspaces();
+  if (_jointTrajectory.joint_names.size() != numControlledDofs)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledDofs << ", got "
+        << _jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  if (_jointTrajectory.points.size() < 2)
+  {
+    throw std::invalid_argument{
+      "Trajectory must contain at least two or more waypoints."};
+  }
+
+  // Check whether joint names match between _space and _jointTrajectory.
+  for (size_t i = 0; i < _jointTrajectory.joint_names.size(); ++i)
+  {
+    auto joint = _space->getJointSpace(i)->getJoint();
+    if (_jointTrajectory.joint_names[i] != joint->getName())
+    {
+      std::stringstream message;
+      message << "Unknown joint name: expected "
+          << joint->getName() << ", got "
+          << _jointTrajectory.joint_names[i] << ".";
+      throw std::invalid_argument{message.str()};
+    }
+  }
```

Comment:
> If we can't assume that they are given in the same order, how do we build the initial mapping? Should we compare each `_jointTrajectory.joint_names[i]` with all joint names in `_space->getJointSpace(j)->getJoint()->getName()`?

That was my idea. None of these solutions are perfect, but I trust the names to be correct more than I trust the indices to be.

> One more question: is it `joint_name` or `dof_name` that we get from `_jointTrajectory.joint_names[]` when we get it from actual HERB?

There is no correct answer here, since the `JointTrajectory` message has no distinction between joints and DOFs. I would strongly prefer `Joint`, since that: (1) it seems that `DegreeOfFreedom` names are typically (but not always) constructed from their `Joint`'s name, (2) URDF only provides a mechanism for specifying joint names, and (3) it semantically matches the `joint_name` field in the message.

> (Assuming single dofs. Are they the same?)

They are not guaranteed to be the same in DART, but I believe they are by default. @jslee02 Is that correct?